ansible_collections/arista/avd/plugins/plugin_utils/schema/avddataconverter.py

```diff
@@ -1,238 +1,166 @@
-                except Exception:
-                    # Ignore errors and return original
-                    converted_instance = instance
-                    pass
-                return converted_instance
-    return instance
-
-
-def _is_float(validator, instance):
-    return isinstance(instance, float)
-
-
-"""
-AvdDataConverter is used to convert AVD Data Types based on schema options.
-We have extra type checkers not covered by the AVD_META_SCHEMA (array, boolean etc)
-since the same TypeChecker is used by the converters themselves.
-"""
-if JSONSCHEMA_IMPORT_ERROR or DEEPMERGE_IMPORT_ERROR:
-    AvdDataConverter = None
-else:
-    AvdDataConverter = jsonschema.validators.create(
-        meta_schema=AVD_META_SCHEMA,
-        validators={"$ref": _ref, "items": _items, "keys": _keys},
-        type_checker=jsonschema.TypeChecker(
-            {
-                "any": jsonschema._types.is_any,
-                "array": jsonschema._types.is_array,
-                "boolean": jsonschema._types.is_bool,
-                "integer": jsonschema._types.is_integer,
-                "object": jsonschema._types.is_object,
-                "null": jsonschema._types.is_null,
-                "None": jsonschema._types.is_null,
-                "number": jsonschema._types.is_number,
-                "string": jsonschema._types.is_string,
-                "dict": jsonschema._types.is_object,
-                "str": jsonschema._types.is_string,
-                "bool": jsonschema._types.is_bool,
-                "list": jsonschema._types.is_array,
-                "int": jsonschema._types.is_integer,
-                "float": _is_float,
-            }
-        ),
-    )
+    def __init__(self, avdschema):
+        self._avdschema = avdschema
+
+        self.converters = {
+            "items": self.convert_items,
+            "keys": self.convert_keys,
+        }
+
+    def convert_data(self, data, schema: dict = None, path: list = None) -> Generator:
+        if schema is None:
+            # Get fully resolved schema (where all $ref has been expanded recursively)
+            # Performs inplace update of the argument so we give an empty dict.
+            # By default it will resolve the full schema
+            schema = {}
+            resolve_errors = self._avdschema.resolve(schema)
+            for resolve_error in resolve_errors:
+                if isinstance(resolve_error, Exception):
+                    raise AristaAvdError(resolve_error)
+
+        if path is None:
+            path = []
+
+        for key, converter in self.converters.items():
+            if key not in schema:
+                # Ignore keys not in schema
+                continue
+
+            # Converters will do inplace update of data. Any returns will be yielded conversion messages.
+            yield from converter(schema[key], data, schema, path)
+
+    def convert_keys(self, keys: dict, data: dict, schema: dict, path: list):
```

Comment:
Probably worth saying what is the format for `path` - a list of strings that give a path in the schema if I get it right?


Missing return type

Reply:
There is no return type.

Reply:
technically there is always a return type but ok - https://stackoverflow.com/questions/42227288/pythons-pep-484-type-annotation-for-generator-expression